src/ucp/proto/proto_init.c

```diff
@@ -698,66 +458,67 @@ ucp_proto_common_check_mem_access(const ucp_proto_common_init_params_t *params)
-                                                   tl_perf_node, reg_md_map, 0,
-                                                   0, caps);
-        if (status != UCS_OK) {
-            return status;
-        }
+    if (ucp_proto_common_check_mem_access(params)) {
+        ucs_assert(tl_perf->max_frag >= params->hdr_size);
+        range_end = ucs_min(params->max_length,
+                            tl_perf->max_frag - params->hdr_size);
+    } else {
+        /* If memory access is not possible, support only empty message */
+        range_end = 0;
     }
 
-    /* Get fragment size */
-    ucs_assert(tl_perf->max_frag >= params->hdr_size);
-    range_end = ucs_min(params->max_length,
-                        tl_perf->max_frag - params->hdr_size);
-    if ((range_end == 0) || !ucp_proto_common_check_mem_access(params)) {
-        /* Return UNSUPPORTED if protocol cannot be used on any range */
-        return (caps->min_length == 0) ? UCS_OK : UCS_ERR_UNSUPPORTED;
+    if (range_end < tl_perf->min_length) {
+        return UCS_ERR_UNSUPPORTED;
     }
 
-    ucs_assertv_always(range_end >= caps->min_length,
-                       "range_end=%zu caps->min_length=%zu",
-                       range_end, caps->min_length);
-
-    /* Add ranges representing sending single fragment */
-    status = ucp_proto_init_single_frag_ranges(params, tl_perf, tl_perf_node,
-                                               reg_md_map, caps->min_length,
-                                               range_end, caps);
+    status = ucp_proto_perf_create(perf_name, &perf);
     if (status != UCS_OK) {
         return status;
     }
 
-    /* Append range representing sending rest of the fragments, if range_end is
-       not the max length and the protocol supports fragmentation */
-    if ((range_end < params->max_length) &&
-        !(params->flags & UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG)) {
-        ucp_proto_common_add_ppln_range(caps,
-                                        &caps->ranges[caps->num_ranges - 1],
-                                        params->max_length);
+    status = ucp_proto_init_add_tl_perf(params, tl_perf, tl_perf_node,
+                                        range_start, range_end, perf);
+    if (status != UCS_OK) {
+        goto out;
     }
 
-    return UCS_OK;
+    if (range_end > 0) {
+        /* Add buffer copy/register cost for non-empty messages */
+        status = ucp_proto_init_add_buffer_perf(params, ucs_max(1, range_start),
+                                                range_end, reg_md_map, perf);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        /* Add range that represents sending many fragments */
+        if ((range_end < params->max_length) &&
+            !(params->flags & UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG)) {
+            frag_seg = ucp_proto_perf_segment_last(perf);
+            status   = ucp_perf_add_ppln(perf, frag_seg, params->max_length);
```

Comment:
We can, but for RNDV pipeline we still need to find fragment segment on the outer level since we define size of fragment as `ucp_proto_perf_segment_end(frag_seg)`.